src/main/com/aca/homework/week8/product/repository/InMemoryProductRepository.java

```diff
@@ -0,0 +1,43 @@
+package com.aca.homework.week8.product.repository;
+
+public class InMemoryProductRepository implements ProductRepository {
+    private final Product[] products = new Product[10000];
+    private int numOfProducts;
+
+    @Override
+    public void save(Product product) {
+        if (product == null) {
+            throw new IllegalArgumentException("Provided product is null.");
+        }
+        if (findById(product.getId()) != null) {
+            throw new ProductAlreadyExistsException("Product with id = " + product.getId() + " already exists.");
```

Comment:
please move the exception message template to the exception constructor and just pass id as a constructor argument